### ISSUE TITLE
System.IndexOutOfRangeException: Index was outside the bounds of the array.

### ISSUE THREAD

**Issue by nberardi** [ericsink/SQLitePCL.raw#57]
We are investigating an issue where the `sqlite3_stmt` is not thread safe and throws `IndexOutOfRangeExceptions` when put under pressure. Do you have any insight on what may be going on?  Would something like ConcurrentDictionary help out in this situation? 

```csharp
System.IndexOutOfRangeException: Index was outside the bounds of the array.
   at System.Collections.Generic.Dictionary`2.Insert(TKey key, TValue value, Boolean add)
   at System.Collections.Generic.Dictionary`2.set_Item(TKey key, TValue value)
   at SQLitePCL.sqlite3.add_stmt(sqlite3_stmt stmt)
   at SQLitePCL.sqlite3_stmt..ctor(IntPtr p, sqlite3 db)
   at SQLitePCL.raw.sqlite3_prepare_v2(sqlite3 db, String sql, sqlite3_stmt& stmt, String& tail)
   at SQLitePCL.raw.sqlite3_prepare_v2(sqlite3 db, String sql, sqlite3_stmt& stmt)
   at SQLite.SQLite3.Prepare2(sqlite3 db, String query) in C:\projects\CloudDriveDesktop\src\Amazon.CloudDrive.Shared\Database\SQLite.cs:line 3265
   at SQLite.SQLiteCommand.Prepare() in C:\projects\CloudDriveDesktop\src\Amazon.CloudDrive.Shared\Database\SQLite.cs:line 2261
   at SQLite.SQLiteCommand.<ExecuteDeferredQuery>d__12`1.MoveNext() in C:\projects\CloudDriveDesktop\src\Amazon.CloudDrive.Shared\Database\SQLite.cs:line 2182
   at System.Collections.Generic.List`1..ctor(IEnumerable`1 collection)
   at System.Linq.Enumerable.ToList[TSource](IEnumerable`1 source)
   at SQLite.SQLiteCommand.ExecuteQuery[T]() in C:\projects\CloudDriveDesktop\src\Amazon.CloudDrive.Shared\Database\SQLite.cs:line 2151
   at SQLite.TableQuery`1.GetEnumerator() in C:\projects\CloudDriveDesktop\src\Amazon.CloudDrive.Shared\Database\SQLite.cs:line 2952
   at System.Collections.Generic.List`1..ctor(IEnumerable`1 collection)
   at System.Linq.Enumerable.ToList[TSource](IEnumerable`1 source)
   at SQLite.TableQuery`1.FirstOrDefault() in C:\projects\CloudDriveDesktop\src\Amazon.CloudDrive.Shared\Database\SQLite.cs:line 2971
```
```

**Comment by ericsink** [comment]
On first glance, that does look suspicious, like it should have gotten the same ConcurrentDictionary changes we did for the hooks.  I'll investigate further.

**Comment by bordoley** [comment]
@nberardi sqlite_stmt isn't threadsafe, nor is a sqlite3 connection. Are you using a lock or other concurrency mechanism to ensure that you don't access a given connection or any of its statements from more than one thread at a time?

I'm not sure it makes sense to use concurrent dictionary within sqlite3.add_stmt, since the underlying connection is not safe for concurrent usage. The change made sense for the hooks, since those dictionaries were mapping connections to hooks. That change was designed to prevent contention issues  when accessing multiple database connections concurrently, which is safe.

**Comment by nberardi** [comment]
@bordoley we are using locks for write statements.  And handling BUSY responses with retries.  However we aren't locking read statements.  However this isn't related to the `sqlite3_stmt` being threadsafe or not, but the managing dictionary in the `sqlite3` class not being thread safe.

**Comment by bordoley** [comment]
You need to lock reads as well. You need to use the same unit of concurrency for a database and all its statements. For instance you can't use two statements from a single database concurrently.

**Comment by bordoley** [comment]
Note: Its worth reading https://www.sqlite.org/threadsafe.html

Using a given database and its statements from more than one thread at a time within SQLite itself is only supported in Serialized mode. In this case all SQLite is doing is adding the locks itself instead of let you own the locking strategy. 

There was another issue regarding support for serialize mode, but as I recall the decision at the time was that you should avoid concurrent access to a sqlite database and its statements.

**Comment by nberardi** [comment]
>> For instance you can't use two statements from a single database concurrently.

If that is true, then there is no need for the Dictionary in the `sqlite3` class.  The Dictionary is managing the `sqlite3_stmt` objects according to pointers.  However, I haven't run into any issues with reads or writes, outside of BUSY responses, which I handle with retires, because all my writes are being performed in transactions.  

Apart from if SQLite is thread-safe or not.  The error I am seeing is directly related to two threads trying to add statements to a dictionary at the same time.  If this worked correctly, the BUSY retries should handle any other situation that comes up.

**Comment by bordoley** [comment]
The dictionary is used to support sqlite3_next_stmt.

**Comment by paulcbetts** [comment]
Even if you get this to not crash, accessing sqlite3 from >1 thread will make it super slow. It's better to queue all your operations then have a single thread dispatch them

**Comment by ericsink** [comment]
OK, so I am tempted to try a ConcurrentDictionary in this case, but only because I think it would be interesting to know whether it gets around the problem @nberardi is having.

As @bordoley observed (and I initially did not), this problem can only happen if you are using a sqlite3 database handle from multiple threads.

And, as @paulcbetts commented, this is always a horrible idea, whether you are using SQLitePCL.raw or not.

And if I put in a ConcurrentDictionary, it will cause a perf hit for everybody, which is bad.

Especially because this particular dictionary is nearly pointless.  It exists only to support sqlite3_next_stmt().

Rock.  Hard place.

**Comment by nberardi** [comment]
I am running a build with ConcurrentDictionary. I will report my findings after putting it through stress tests that have yielded this error reliably. 

--
Nick Berardi
(484) 302-0125
Sent on the go from my phone.

**Comment by nberardi** [comment]
@ericsink I don't think @bordoley analysis is correct, because it is happening on adding to the dictionary not removing.  Which means two threads are trying to expand the internal array of the dictionary at the same time.  And one looses because Dictionary<,> is not thread safe.  Regardless of how it is used to get stmt's later on, when adding to the dictionary you cannot assume thread safety just because everybody has an unwritten rule to put everything that executes in a lock statement.  

Bugs happen, and in my case we inherited code where we didn't have locks around everything. As a library you should just be a pass through to the effects of sqlite3, if multiple requests slows sqlite3 down, so be it.  But the wrapper library it self shouldn't crash.  

Lastly to @paulcbetts and @bordoley point about ConcurrentDictionary taking a performance hit.  This is only true if multiple threads are accessing it.  If a single thread is accessing it like @paulcbetts says is the right way, it should have zero effect on the performance of ConcurrentDictionary vs Dictionary.  ConcurrentDictionary with the current design will only provide a performance hit to the libraries that aren't operating in thread safety.

I wasn't aware until @bordoley brought it up that we are using the Dictionary as a queue, if this is the cases there are more elegant solutions to this, that we can provide a large initial buffer of space for.

**Comment by ericsink** [comment]
@nberardi said: "As a library you should just be a pass through to the effects of sqlite3, if multiple requests slows sqlite3 down, so be it. But the wrapper library it self shouldn't crash."

See now you've gone and made everything complicated.  I was close to agreeing with you on this issue until you said THAT.

Your reasoning above *sounds* good, but in practice, I am concerned that you (people who want to use one sqlite3 connection handle from multiple threads) are a mouse who, if I give you a cookie, will just want a glass of milk.

In other words:  If the only thing necessary to make your kind happy is to put a ConcurrentDictionary in that one spot, sure, I might actually want to do that.

But I wonder if your reasoning can be paraphrased as:  "Eric must make any and all fixes to allow SQLitePCL.raw to fully support use of a sqlite3 handle from multiple threads, not matter what those fixes are, and no matter what the consequences are."

That's a commitment I don't want to make until more is known about the rest of the steps down that path.

**Comment by nberardi** [comment]
Well if you were agreeing with me, then scratch it out.  :+1:   In all seriousness though...

> "Eric must make any and all fixes to allow SQLitePCL.raw to fully support use of a sqlite3 handle from multiple threads, not matter what those fixes are, and no matter what the consequences are."

@ericsink I may have misspoke a little in my statement. I am not saying that you should support multi-threading. I am saying the library shouldn't crash except when crashing based on something from sqlite3 happening.  I am all for reducing your work load.

I dug into `sqlite3_next_stmt` and it is not used by SQLite-net.  Which means that the Dictionary is overhead for me and whoever else uses your library with SQLite-net or the derivative builds of it. I am now wondering if the `sqlite3` object is the right place to have this pointer to reference storage.

**Comment by ericsink** [comment]
Sometimes I wish I had never added support for sqlite3_next_stmt(), but now...

I'm reluctant to remove support for sqlite3_next_stmt() as it would break apps that use it.

The main question remaining is whether ConcurrentDictionary in that spot does in fact fix your problem.  I'll wait to hear the result of your stress testing.

**Comment by nberardi** [comment]
So far we are looking good, it will be through our automation tests on Monday.

**Comment by nberardi** [comment]
Just as a follow up, I added a pull request related to the change that I am running through my tests in regards to ConcurrentDictionary.  I also added a flag that could switch off adding and removing values from the dictionary if so desired to reduce any overhead associated with sqlite3_next_stmt.

**Comment by nberardi** [comment]
I haven't seen a repeat of this issue after introducing the `ConcurrentDictionary`.  I would say that solved the problem that caused this exception.

**Comment by ericsink** [comment]
FWIW, Nuget package 0.8.3-pre1 contains the change to use a ConcurrentDictionary for this case.

**Comment by ericsink** [comment]
0.8.3 is on nuget and contains the fix for this.
